plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-orders-controller.php

```diff
@@ -170,6 +170,46 @@ protected function prepare_object_for_database( $request, $creating = false ) {
 		return apply_filters( "woocommerce_rest_pre_insert_{$this->post_type}_object", $order, $request, $creating );
 	}
 
+	/**
+	 * Wrapper method to remove order items.
+	 * When updating, the item ID provided is checked to ensure it is associated
+	 * with the order.
+	 *
+	 * @param WC_Order $order     The order to remove the item from.
+	 * @param string   $item_type The item type (from the request, not from the item, e.g. 'line_items' rather than 'line_item').
+	 * @param int      $item_id   The ID of the item to remove.
+	 *
+	 * @return void
+	 * @throws WC_REST_Exception If item ID is not associated with order.
+	 */
+	protected function remove_item( WC_Order $order, string $item_type, int $item_id ) {
```

Comment:
- Not sure we really have established rules on this, and definitely not a deal breaker, but should we start with private visibility? When it's protected, it's still arguably part of the public API surface (since a third party class can access it through inheritance). By reducing visibility, if we later wish to remove or refactor, we'd have fewer back-compat concerns.
- Declare the return type of `void`?

Reply:
> should we start with private visibility?

Maybe, although I think there are a couple of reasons not to:

* The matching [`set_item`](https://github.com/woocommerce/woocommerce/blob/0468bdbe82b2ebb0d3c8ec04bdb7c840b3e7de3b/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-orders-v2-controller.php#L1014-L1064) method has protected visibility, so I think it makes sense to do the same here for consistency.
* The orders controller class gets extended by some other endpoint controllers (notably refunds), and if those controllers use the inherited `prepare_object_for_database` instead of overwriting it, they'll need access to `remove_item` as well.

> Declare the return type of void?

Good point. Added in f9ed8f9282

Reply:
Totally academic at this point, so feel free to ignore (I'm continuing the discussion out of interest, mostly):

> so I think it makes sense to do the same here for consistency

This bit makes sense...

> if those controllers use the inherited prepare_object_for_database instead of overwriting it, they'll need access to remove_item as well.

I don't quite understand this bit, though. If they call the inherited `prepare_object_for_database` and *that* calls **private** method `remove_item`, that's completely fine. Or do you mean a call to `prepare_object_for_database` will tend to be followed up with a direct call to `remove_item`?

Reply:
Ah, see, this is me being confused about how class inheritance works. I assumed that `$this` refers to the actual class being used, even if invoked from a method in the parent class. But it appears that I'm wrong:

https://3v4l.org/Mp12f#v8.3.10

Reply:
Still, I think the consistency aspect (consistency with its peer method, `set_item`) is totally valid.